clash_lib/src/proxy/utils/test_utils/docker_runner.rs

```diff
@@ -0,0 +1,237 @@
+//! This example will run a non-interactive command inside the container using `docker exec`
+
+use std::collections::HashMap;
+
+use bollard::container::{Config, RemoveContainerOptions};
+use bollard::secret::{HostConfig, Mount, PortBinding};
+use bollard::Docker;
+
+use bollard::exec::{CreateExecOptions, StartExecResults};
+use bollard::image::CreateImageOptions;
+
+use anyhow::Result;
+use futures::{Future, StreamExt, TryStreamExt};
+
+pub struct DockerTestRunner {
+    instance: Docker,
+    id: String,
+}
+
+impl DockerTestRunner {
+    pub async fn new<'a>(
+        image_conf: Option<CreateImageOptions<'a, String>>,
+        container_conf: Config<String>,
+    ) -> Result<Self> {
+        let docker: Docker = Docker::connect_with_socket_defaults()?;
+
+        docker
+            .create_image(image_conf, None, None)
+            .try_collect::<Vec<_>>()
+            .await?;
+
+        let id = docker
+            .create_container::<&str, String>(None, container_conf)
+            .await?
+            .id;
+        docker.start_container::<String>(&id, None).await?;
+        Ok(Self {
+            instance: docker,
+            id,
+        })
+    }
+
+    #[allow(unused)]
+    pub async fn exec(&self, cmd: Vec<&str>) -> anyhow::Result<()> {
+        // non interactive
+        let exec = self
+            .instance
+            .create_exec(
+                &self.id,
+                CreateExecOptions {
+                    attach_stdout: Some(true),
+                    attach_stderr: Some(true),
+                    cmd: Some(cmd),
+                    ..Default::default()
+                },
+            )
+            .await?
+            .id;
+        if let StartExecResults::Attached { mut output, .. } =
+            self.instance.start_exec(&exec, None).await?
+        {
+            while let Some(Ok(msg)) = output.next().await {
+                print!("{msg}");
+            }
+            return Ok(());
+        } else {
+            anyhow::bail!("failed to execute cmd")
+        }
+    }
+
+    // will make sure the container is cleaned up after the future is finished
+    pub async fn run_and_cleanup(
+        self,
+        f: impl Future<Output = anyhow::Result<()>>,
+    ) -> anyhow::Result<()> {
+        let fut = Box::pin(f);
+        let res = fut.await;
```

Comment:
maybe give this `fut.await` a time out just in case it blocks.

and cargo doesn't seem to kill timeout tasks either https://stackoverflow.com/questions/73131115/unit-test-timeout-in-rust#comment134752168_73133545